### PR TITLE
237/add crawler for bundesfreiwilligendienst.de

### PR DESCRIPTION
**Pull request by DanielKraft**
Crawler implementation for bundesfreiwilligendienst.de

Can be tested with the following commands:

```
$ cd etl
$ python (or python3)
>>> from shared.utils import write_data_to_json
>>> write_data_to_json('data_extraction/data/test.json', [])

>>> from data_extraction.scraper.bundesfreiwilligendienst import BundesFreiwilligendienst as scr
>>> i = scr('test')
>>> i.add_urls()

>>> import random
>>> for j in range(0, 20):
...    i.crawl(i.urls[j], random.randrange(len(i.urls)))
```

### CODE REVIEW

**Review by FiedlerJan — CHANGES_REQUESTED** [review]
While executing 'add_urls()' i get the following error.

```
Fetched 100 URLs from https://www.bundesfreiwilligendienst.de [35/121]
Fetched 100 URLs from https://www.bundesfreiwilligendienst.de [36/121]
Traceback (most recent call last):
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.9/http/client.py", line 276, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/util/retry.py", line 531, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.9/http/client.py", line 276, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/janf/git/einander-helfen/etl/data_extraction/scraper/bundesfreiwilligendienst.py", line 139, in add_urls
    response, session = self.soupify_post_session(next_page_url, {
  File "/home/janf/git/einander-helfen/etl/data_extraction/Scraper.py", line 120, in soupify_post_session
    res = session.post(url, data=form_data)
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/requests/sessions.py", line 590, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/home/janf/git/einander-helfen/etl/venv/lib/python3.9/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

**Comment by DanielKraft** [comment]
Can you run it again on your end, because on my end add_urls() runs without any problems.

File: etl/data_extraction/scraper/bundesfreiwilligendienst.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            'target_group': None,
+            'prerequisites': None,
+            'language_skills': None,
+            'timing': None,
+            'effort': None,
+            'opportunities': None,
+            'organization': None,
+            'contact': contact.decode_contents().strip() if contact is not None else None,
+            'link': url or None,
+            'source': self.base_url,
```
Comment (jonas-hda) [inline etl/data_extraction/scraper/bundesfreiwilligendienst.py]: change source to "www.bundesfreiwilligendienst.de"

**Review by jonas-hda — CHANGES_REQUESTED** [review]

File: etl/data_extraction/scraper/bundesfreiwilligendienst.py (hunk 1 above)
Comment (jonas-hda) [reply etl/data_extraction/scraper/bundesfreiwilligendienst.py]: 1 commit to late

**Comment by FiedlerJan** [comment]
> 
> 
> Can you run it again on your end, because on my end add_urls() runs without any problems.

I could not reproduce the problem either.

**Review by FiedlerJan — APPROVED** [review]

**Review by jonas-hda — APPROVED** [review]

**Pull request merged**